File: 0112-path-sum/0112-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function(root, targetSum) {
    // Handle the base case(s), usually an empty tree (node = null) is a base case.
    // Do some logic for the current node
    // Recursively call on the current node's children
    // Return the answer
    if (!root) {
        return false;
    }
    targetSum -= root.val;
    if (!root.left && !root.right) {
        return targetSum === 0;
    }
    return hasPathSum(root.left, targetSum) || hasPathSum(root.right, targetSum);
};

File: 0112-path-sum/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int target;
    public boolean hasPathSum(TreeNode root, int targetSum) {
        target = targetSum;
        return pathSum(root, 0);
    }
    public boolean pathSum(TreeNode node, int currSum) {
        if (node == null) {
            return false;
        }
        if (node.left == null & node.right == null) {
            return (currSum + node.val) == target;
        }
        currSum += node.val;
        return pathSum(node.left, currSum) || pathSum(node.right, currSum);
    }
}